### ISSUE TITLE
Incorrect error message when accessing private field of union

### ISSUE THREAD

**Issue by gnzlbg** [rust-lang/rust#63976]
```rust
#[derive(Copy, Clone)]
pub struct B;

mod a {
    pub union U {
        pub a: i32,
        b: crate::B,
    }
}

fn new(a: i32) -> B {
    unsafe { a::U { a }.b }
}
```

([Playground](https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=d461d59d40a77187aa1f38e1861f0801))

Errors:

```
    Compiling playground v0.0.1 (/playground)
error[E0616]: field `b` of struct `a::U` is private
  --> src/lib.rs:12:14
   |
12 |     unsafe { a::U { a }.b }
   |              ^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0616`.
error: Could not compile `playground`.

To learn more, run the command again with --verbose.
```

Notice how it says "struct" when it should say "union".

cc @estebank this looks like low-hanging fruit

**Comment by sam09** [comment]
Can I pick this up if no-one is working on this?

**Comment by sam09** [comment]
Thanks @estebank .
Newbie question, but how do I test this while development? Do I build the stage1 compiler and then write a sample test?

**Comment by sam09** [comment]
https://github.com/rust-lang/rust/pull/63982

**Comment by estebank** [comment]
@sam09 `./x.py build src/test/ui/PATH_TO_FILE -i --stage 1 --keep-stage 1` is the way to get the fastest iteration time. `--keep-stage 1` is theoretically suspect but should affect you on what you're working on and `-i` is marginally slower on first compile but much faster on subsequent compiles. You rarely (if ever) need `--stage 2`.
